src/main/drivers/rangefinder/rangefinder_lidartf.c

```diff
@@ -0,0 +1,276 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#ifdef USE_RANGEFINDER_TF
+
+#include "build/debug.h"
+#include "build/build_config.h"
+
+#include "io/serial.h"
+
+#include "drivers/time.h"
+#include "drivers/rangefinder/rangefinder.h"
+#include "drivers/rangefinder/rangefinder_lidartf.h"
+
+#define TF_DEVTYPE_NONE 0
+#define TF_DEVTYPE_MINI 1
+#define TF_DEVTYPE_02   2
+
+static uint8_t tfDevtype = TF_DEVTYPE_NONE;
+
+#define TF_FRAME_LENGTH    6             // Excluding sync bytes (0x59) x 2 and checksum
+#define TF_FRAME_SYNC_BYTE 0x59
+#define TF_TIMEOUT_MS      (100 * 2)
+
+//
+// Benewake TFmini frame format
+// Byte Off Description
+// 1    -   SYNC
+// 2    -   SYNC
+// 3    0   Measured distance (LSB)
+// 4    1   Measured distance (MSB)
+// 5    2   Signal strength (LSB)
+// 6    3   Signal strength (MSB)
+// 7    4   Integral time
+// 8    5   Reserved
+// 9    -   Checksum (Unsigned 8-bit sum of bytes 0~7)
+//
+// Credibility
+// 1. If distance is 12m (1200cm), then OoR.
+//
+#define TF_MINI_FRAME_INTEGRAL_TIME 4
+
+//
+// Benewake TF02 frame format (From SJ-GU-TF02-01 Version: A01)
+// Byte Off Description
+// 1    -   SYNC
+// 2    -   SYNC
+// 3    0   Measured distance (LSB)
+// 4    1   Measured distance (MSB)
+// 5    2   Signal strength (LSB)
+// 6    3   Signal strength (MSB)
+// 7    4   SIG (Reliability in 1~8, less than 7 is unreliable)
+// 8    5   TIME (Exposure time, 3 or 6)
+// 9    -   Checksum (Unsigned 8-bit sum of bytes 0~7)
+//
+// Credibility
+// 1. If SIG is less than 7, unreliable
+// 2. If distance is 22m (2200cm), then OoR.
+//
+#define TF_02_FRAME_SIG 4
+
+// Maximum ratings
+
+#define TF_MINI_RANGE_MIN 40
+#define TF_MINI_RANGE_MAX 1200
+
+#define TF_02_RANGE_MIN 40
+#define TF_02_RANGE_MAX 2200
+
+#define TF_DETECTION_CONE_DECIDEGREES 900
+
+static serialPort_t *tfSerialPort = NULL;
+
+typedef enum {
+    TF_FRAME_STATE_WAIT_START1,
+    TF_FRAME_STATE_WAIT_START2,
+    TF_FRAME_STATE_READING_PAYLOAD,
+    TF_FRAME_STATE_WAIT_CKSUM,
+} tfFrameState_e;
+
+static tfFrameState_e tfFrameState;
+static uint8_t tfFrame[TF_FRAME_LENGTH];
+static uint8_t tfReceivePosition;
+
+// TFmini
+// Command for 100Hz sampling (10msec interval)
+// At 100Hz scheduling, skew will cause 10msec delay at the most.
+static uint8_t tfCmdTFmini[] = { 0x42, 0x57, 0x02, 0x00, 0x00, 0x00, 0x01, 0x06 };
+
+// TF02
+// Same as TFmini for now..
+static uint8_t tfCmdTF02[] = { 0x42, 0x57, 0x02, 0x00, 0x00, 0x00, 0x01, 0x06 };
+
+static int32_t lidarTFValue;
+static uint16_t lidarTFerrors = 0;
+
+static void lidarTFSendCommand(void)
+{
+    switch (tfDevtype) {
+    case TF_DEVTYPE_MINI:
+        serialWriteBuf(tfSerialPort, tfCmdTFmini, sizeof(tfCmdTFmini));
```

Comment:
Is talking to serial port in a device driver a good idea?

Reply:
@digitalentity What do you suggest?

- Move `rangefinder_lidartf.c` to `io` or `io/rangefinder`.
- Open the serial port in `sensors/rangefinder.c` and pass the `serialPort_t` to `detect` or `init`.

Reply:
@jflyper check the opflow_virtual implementation
The driver (gateway to IO code): https://github.com/iNavFlight/inav/blob/66da4ab0129962b967e5d05db1d021bc7840cc17/src/main/drivers/opflow/opflow_virtual.c

Actual interface code (operating at IO level): https://github.com/iNavFlight/inav/blob/66da4ab0129962b967e5d05db1d021bc7840cc17/src/main/io/opflow_cxof.c

Sensor detection: https://github.com/iNavFlight/inav/blob/66da4ab0129962b967e5d05db1d021bc7840cc17/src/main/sensors/opflow.c#L94

Took inspiration from VTX control implementation. IO code implements APIs and publishes a VTable. Driver code receives a vtable when attempting detection.

Reply:
Ah...